alibi_detect/od/_gmm.py

```diff
@@ -0,0 +1,214 @@
+        backend
+            Backend used for outlier detection. Defaults to ``'pytorch'``. Options are ``'pytorch'`` and ``'sklearn'``.
+        device
+            Device type used. The default tries to use the GPU and falls back on CPU if needed. Can be specified by
+            passing either ``'cuda'``, ``'gpu'`` or ``'cpu'``.
+
+        Raises
+        ------
+        NotImplementedError
+            If choice of `backend` is not implemented.
+        """
+        super().__init__()
+
+        backend_str: str = backend.lower()
+        BackendValidator(
+            backend_options={'pytorch': ['pytorch'], 'sklearn': ['sklearn']},
+            construct_name=self.__class__.__name__
+        ).verify_backend(backend_str)
+
+        backend_cls = backends[backend]
+        args: Dict[str, Any] = {'n_components': n_components}
+        if backend == 'pytorch':
+            args['device'] = device
+        self.backend = backend_cls(**args)
+
+    def fit(
+        self,
+        x_ref: np.ndarray,
+        optimizer: Optional[str] = 'Adam',
+        learning_rate: float = 0.1,
+        batch_size: Optional[int] = None,
+        epochs: Optional[int] = None,
+        tol: float = 1e-3,
+        n_init: int = 1,
+        init_params: str = 'kmeans',
+        verbose: int = 0,
+    ) -> None:
+        """Fit the detector on reference data.
+
+        If the ``'pytorch'`` backend is used, the detector is fitted using gradient descent. This is the recommended
+        backend for larger datasets.
+
+        If the ``'sklearn'`` backend is used, the detector is fitted using the EM algorithm. The ``'sklearn'``
+        backend is recommended for smaller datasets. For more information on the EM algorithm and the sklearn Gaussian
+        Mixture Model, see `here <https://scikit-learn.org/stable/modules/generated/sklearn.mixture.GaussianMixture.html#sklearn.mixture.GaussianMixture>`_.  # noqa: E501
+
+        Parameters
+        ----------
+        x_ref
+            Reference data used to fit the detector.
+        optimizer
+            Optimizer used to fit the detector. Only used if the ``'pytorch'`` backend is used. Defaults to ``'Adam'``.
+        learning_rate
+            Learning rate used to fit the detector. Only used if the ``'pytorch'`` backend is used. Defaults to ``0.1``.
+        batch_size
+            Batch size used to fit the detector. Only used if the ``'pytorch'`` backend is used. Defaults to ``None``.
+            If ``None``, the entire dataset is used in each epoch.
+        epochs
+            Number of epochs used to fit the detector. Used for both the ``'pytorch'`` and ``'sklearn'`` backends.
+            If the backend is ``'sklearn'``, the detector is fitted using the EM algorithm and the number of epochs
+            defualts to ``10``. If the backend is ``'pytorch'``, the detector is fitted using gradient descent and
+            the number of epochs defaults to ``100``.
+        tol
```

Comment:
Can we also implement tol for the pytorch variant? Would only be a few extra lines of code (see OCSVM demo)

Reply:
And then rename `epochs` to `max_epochs` to align with sklearn's `max_iter`